### PR TITLE
Implement Comment command for users to leave comments

### PR DESCRIPTION
**Pull request by chiralcentre**
The following files were modified or created:
        new file:   src/main/java/seedu/address/logic/commands/CommentCommand.java
        modified:   src/main/java/seedu/address/logic/commands/EditCommand.java
        modified:   src/main/java/seedu/address/logic/parser/AddCommandParser.java
        modified:   src/main/java/seedu/address/logic/parser/AddressBookParser.java
        modified:   src/main/java/seedu/address/logic/parser/CliSyntax.java
        new file:   src/main/java/seedu/address/logic/parser/CommentCommandParser.java
        new file:   src/main/java/seedu/address/model/person/Comment.java
        modified:   src/main/java/seedu/address/model/person/Person.java
        modified:   src/main/java/seedu/address/storage/JsonAdaptedPerson.java
        modified:   src/main/java/seedu/address/ui/PersonCard.java
        modified:   src/main/resources/view/PersonListCard.fxml
        modified:   src/test/data/JsonSerializableAddressBookTest/duplicatePersonAddressBook.json
        modified:   src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json
        modified:   src/test/java/seedu/address/logic/commands/CommandTestUtil.java
        new file:   src/test/java/seedu/address/logic/commands/CommentCommandTest.java
        modified:   src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java
        modified:   src/test/java/seedu/address/testutil/PersonBuilder.java
        modified:   src/test/java/seedu/address/testutil/TypicalPersons.java

### CODE REVIEW

**Review by logical-1985516 — APPROVED** [review]
Other than a few nits, LGTM. Well done!

File: src/main/java/seedu/address/logic/commands/CommentCommand.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return new CommandResult(generateSuccessMessage(editedPerson));
+    }
+
+    /**
+     * Generates a command execution success message based on whether
+     * the remark is added to or removed from
+     * {@code personToEdit}.
+     */
+    private String generateSuccessMessage(Person personToEdit) {
+        String message = !comment.value.isEmpty() ? MESSAGE_ADD_COMMENT_SUCCESS : MESSAGE_DELETE_COMMENT_SUCCESS;
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/main/java/seedu/address/logic/commands/CommentCommand.java:74]: I think this is correct; when the comment value is not empty the message should indicate success

File: src/main/java/seedu/address/logic/commands/CommentCommand.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+public class CommentCommand extends Command {
+
+    public static final String COMMAND_WORD = "comment";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Edits the comment of the person identified "
+            + "by the index number used in the last person listing. "
+            + "Existing comment will be overwritten by the input.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "m/ [comment]\n"
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/main/java/seedu/address/logic/commands/CommentCommand.java:29]: Good catch, will fix

File: src/main/java/seedu/address/logic/commands/CommentCommand.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /**
+     * Generates a command execution success message based on whether
+     * the remark is added to or removed from
+     * {@code personToEdit}.
+     */
+    private String generateSuccessMessage(Person personToEdit) {
+        String message = !comment.value.isEmpty() ? MESSAGE_ADD_COMMENT_SUCCESS : MESSAGE_DELETE_COMMENT_SUCCESS;
+        return String.format(message, personToEdit);
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/main/java/seedu/address/logic/commands/CommentCommand.java:75]: I think "personToEdit" is correct; its possible that the comment operation fails

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 4]
```diff
@@ -11,5 +11,6 @@ public class CliSyntax {
     public static final Prefix PREFIX_EMAIL = new Prefix("e/");
     public static final Prefix PREFIX_ADDRESS = new Prefix("a/");
     public static final Prefix PREFIX_TAG = new Prefix("t/");
+    public static final Prefix PREFIX_COMMENT = new Prefix("m/");
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/main/java/seedu/address/logic/parser/CliSyntax.java:14]: done

File: src/main/java/seedu/address/logic/parser/CommentCommandParser.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    @Override
+    public CommentCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args,
+                PREFIX_COMMENT);
+
+        Index index;
+        try {
+            index = ParserUtil.parseIndex(argMultimap.getPreamble());
+        } catch (IllegalValueException ive) {
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/main/java/seedu/address/logic/parser/CommentCommandParser.java:26]: good point; fixed

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 6]
```diff
@@ -45,6 +48,8 @@ public class CommandTestUtil {
     public static final String EMAIL_DESC_BOB = " " + PREFIX_EMAIL + VALID_EMAIL_BOB;
     public static final String ADDRESS_DESC_AMY = " " + PREFIX_ADDRESS + VALID_ADDRESS_AMY;
     public static final String ADDRESS_DESC_BOB = " " + PREFIX_ADDRESS + VALID_ADDRESS_BOB;
+    public static final String COMMENT_DESC_AMY = " " + PREFIX_COMMENT + VALID_COMMENT_AMY;
+    public static final String COMMENT_DESC_BOB = " " + PREFIX_COMMENT + VALID_COMMENT_BOB;
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/test/java/seedu/address/logic/commands/CommandTestUtil.java:52]: They are used in CommentCommandTest.java

File: src/test/java/seedu/address/logic/commands/CommentCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        assertCommandSuccess(commentCommand, model, expectedMessage, expectedModel);
+    }
+
+    @Test
+    public void execute_deleteCommentUnfilteredList_success() {
+        Person firstPerson = model.getFilteredPersonList().get(INDEX_FIRST_PERSON.getZeroBased());
+        Person editedPerson = new PersonBuilder(firstPerson).withComment("").build();
+
+        CommentCommand commentCommand = new CommentCommand(INDEX_FIRST_PERSON,
+                new Comment(editedPerson.getComment().toString()));
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/test/java/seedu/address/logic/commands/CommentCommandTest.java:56]: There is no difference, I have standardised to value

File: src/test/java/seedu/address/model/person/PersonTest.java [hunk 8]
```diff
@@ -93,7 +93,8 @@ public void equals() {
     @Test
     public void toStringMethod() {
         String expected = Person.class.getCanonicalName() + "{name=" + ALICE.getName() + ", phone=" + ALICE.getPhone()
-                + ", email=" + ALICE.getEmail() + ", address=" + ALICE.getAddress() + ", tags=" + ALICE.getTags() + "}";
+                + ", email=" + ALICE.getEmail() + ", address=" + ALICE.getAddress() + ", tags=" + ALICE.getTags()
+                + ", comment=" + "}";
```
[the remark this answers is not in the document]
Comment (chiralcentre) [reply src/test/java/seedu/address/model/person/PersonTest.java:97]: Good point and I added ALICE.getComment() for consistency. However, note that ALICE's comment is an empty string, and it is not strictly required.

**Review by logical-1985516 — APPROVED** [review]
LGTM!

**Pull request merged**